docs(broker): Describe what Broker does and how to interact with it (#571)
openactive/openactive-test-suite: packages/openactive-broker-microservice/README.md

﻿# openactive-broker-microservice

> Broker Microservice sits in front of a [Booking System](#booking-system-under-test), which is an implementation of the [Open Booking API specification](https://openactive.io/open-booking-api/EditorsDraft/), and provides the following services, to enable Test Suite:

Author: Done

1. [Harvests](#initial-harvest) data from the Booking System's Opportunity and Order RPDE feeds.
    * This data is then [validated](#data-model-validation) and [cached](#opportunity-id-cache) and any [listeners](#two-phase-listeners) are notified.
2. Fetch and parse data from the Booking System's [Dataset Site](https://openactive.io/dataset-api-discovery/EditorsDraft/).

> 3. Set up and maintains [auth credentials](#auth) for access to the Booking System.

Author: Done

It needs to be running, against a Booking System, in order for the [openactive-integration-tests](../openactive-integration-tests/) to run.

## Usage in separate terminal windows

> To run `openactive-broker-microservice` in separate terminal window to `openactive-integration-tests`, from repository root:

Author: I've created a draft issue for this: https://github.com/orgs/openactive/projects/17/views/8?pane=issue&itemId=39029519 and so will not do it in this PR

          Simple Booking Flow and satisfy the criteria `TestOpportunityBookable` */
          "https://acme-fitness.org/api/sellers/1": 129,
          "https://acme-fitness.org/api/sellers/2": 27
        },
        "FacilityUseSlot": {
          // ...
        },
        "IndividualFacilityUseSlot": {
          // ...
        }
      },
      "OpenBookingApprovalFlow": {
        // ...
      }
    },
    "TestOpportunityBookableNonFreeTaxNet": {
      "OpenBookingSimpleFlow": {
        "ScheduledSession": {
          /* Unlike with TestOpporunityBookable, there were no ScheduledSessions
          that matched this criteria, `TestOpportunityBookableNonFreeTaxNet`, so
          instead of showing Opportunity numbers for each Seller, the status
          endpoint shows a summary of why no Opportunities were found. */
          "criteriaErrors": {
            /* There are 892 ScheduledSessions which support Simple Booking Flow
            and that do not satisfy the `TestOpportunityBookableNonFreeTaxNet`
            criteria because they fail this contraint. That is to say that they
            do not have seller tax mode TaxNet. */

>             "Seller Tax Mode Net": 892,

Author: Done

        // The RPDE item ID of the Orphan
        "id": "https://acme-fitness.org/api/scheduled-sessions/1",
        "modified": 1234,
        // The JSON-LD data for the Orphan
        "jsonLd": {
          "@context": "https://openactive.io/",
          "@type": "ScheduledSession",
          "@id": "https://acme-fitness.org/api/scheduled-sessions/1",
          "name": "Yoga",
          "superEvent": "https://acme-fitness.org/api/session-series/1",
          // ...etc
        },
        // The JSON-LD ID of the Orphan
        "jsonLdId": "https://acme-fitness.org/api/scheduled-sessions/1",
        // The anticipated JSON-LD ID of the Orphan's parent
        "jsonLdParentId": "https://acme-fitness.org/api/session-series/1",
      },
      // ... More orphans
    ]
  }
}
```

#### `GET /opportunity-cache/:id`

**Request params**:

* `id`: The ID of the [Child Opportunity](#orphans).

**Response Type**: 🤖 JSON

Get an expanded* [Child Opportunity](#orphans) (e.g. a ScheduledSession or IndividualFacilityUseSlot) from Broker Microservice's cache, by its ID. Useful for debugging issues with the [Booking System](#booking-system-under-test)'s data.

> - *expanded: The Opportunity will have been expanded to include its parent Opportunity, if it has one.

Author: Done

**Request params**:

* `id`: The ID of the [Child Opportunity](#orphans) that is already being listened for.

**Response Type**: 🤖 JSON

Must be called after [`POST /opportunity-listeners/:id`](#post-opportunity-listenersid).

If and when the specified Child Opportunity is updated, this will return the updated Opportunity. This invokes the **Get** phase of the [Opportunity Listener](#two-phase-listeners).

Either this endpoint will:

* If an update was found:
    * return the updated Opportunity.
* If no update is found:
    * eventually timeout.

### `POST /order-listeners/:type/:bookingPartnerIdentifier/:uuid`

**Request params**:

* `type`: `orders` or `order-proposals`.
* `bookingPartnerIdentifier`: Identifier of the Booking Partner whose [Orders Feed](https://openactive.io/open-booking-api/EditorsDraft/#dfn-orders-feed) or OrderProposals Feed to listen to. This is the key of the Booking Partner in the [`bookingPartners` configuration property](#bookingpartners) e.g. `primary`.
* `uuid`: UUID of the Order or OrderProposal to listen for.

**Response Type**: 🤖 JSON

Create an [Order Listener](#two-phase-listeners) to listen for updates to the specified Order or Order Proposal.

> The response contains info that is useful to a human user for debugging, such as the `startingFeedPage`, which is the page of the feed that Broker Microservice will start listening from.

Reviewer: Should a copy of an example response be included here, as with others above?

Author: Done

Assert that an Opportunity matching the Opportunity Criteria, Seller and Opportunity Type from the request body is **not** found in the [Booking System](#booking-system-under-test)'s data. If such an Opportunity is found, the response will be a `404`. Otherwise, the response will be a `204`.

This is used by some [Non-Implemented tests](../openactive-integration-tests/README.md#structure) to ensure that a test feature MUST be implemented if a certain criteria of Opportunity is found in the Booking System's data. e.g. the [`cancellation-window` feature](../openactive-integration-tests/test/features/cancellation/cancellation-window) has a non-implemented test that asserts that there are no Opportunities with cancellation windows defined.

## Browser Automation for Auth Endpoints

Broker Microservice also exposes endpoints created from the `setupBrowserAutomationRoutes(..)` function from [openactive-openid-test-client](../openactive-openid-test-client/).
These endpoints are used by [Integration Tests](../openactive-integration-tests/) for tests which check the [OpenID Connect](https://openid.net/developers/how-connect-works/) authentication flow.

## Concepts

> ### Auth

Reviewer: ```suggestion
### Authentication
```

Author: My understanding is that Broker also manages authorization as well as authentication e.g. it maintains the booking partners' access tokens.

Reviewer: Great point - something I've noticed is an issue in the booking spec too (and have raised https://github.com/openactive/open-booking-api/issues/240). Perhaps "Authentication and Authorisation" could be a good solution here too?

Reviewer: However reflecting on this - in this specific case, is the configuration here more focussed on Authentication than Authorisation?

Broker Microservice sets up and maintains auth credentials for access to the [Booking System](#booking-system-under-test).

[Integration Tests](../openactive-integration-tests/) can also obtain these credentials from [`GET /config`](#get-config) and use them to access the Booking System.

> These credentials are set up according to the auth strategy defined in the [`bookingPartners` configuration property](#bookingpartners).

Reviewer: ```suggestion
These credentials are set up according to the authentication strategy defined in the [`bookingPartners` configuration property](#bookingpartners).
```

### Harvesting Status

Broker Microservice can have one of two **Harvesting Statuses**:

* `harvesting`: Broker Microservice is harvesting or polling the [Booking System](#booking-system-under-test)'s RPDE feeds, which means that it is either in its [Initial Harvest](#initial-harvest) phase or it is polling the feeds to keep up to date.
* `paused`: Broker Microservice is **not** harvesting or polling the [Booking System](#booking-system-under-test)'s RPDE feeds. This means that its data is not up to date with the Booking System's data, and so it is not available for use by [Integration Tests](../openactive-integration-tests/).

### Opportunity ID Cache

Broker Microservice keeps a cache of all the [Opportunities](https://openactive.io/open-booking-api/EditorsDraft/#dfn-opportunity) that it has harvested from the [Booking System](#booking-system-under-test)'s feeds.

This cache is stored in various structures. For the purposes of interfacing with Broker Microservice, the most important one to know about is **Buckets**:

> #### Buckets

Reviewer: ```suggestion
### Buckets
```

Author: This is supposed to be a sub-heading under Opportunity ID Cache

> A **Bucket** is a cache of Opportunity IDs that match a given [Opportunity Criteria](../test-interface-criteria/README.md). When Integration Tests sends a request to get a random Opportunity matching a given criteria, the Opportunity is fetched from these buckets.

Reviewer: ```suggestion
A **Bucket** is a cache of Opportunity IDs that match a given [Opportunity Criteria](../test-interface-criteria/README.md). During harvesting, each Opportunity found in the feeds added to the relevant buckets.  When Integration Tests send a request to get a random Opportunity matching a given criteria, the Opportunity is fetched from these buckets.
```

Author: Done

### Opportunity Locks

When running tests in [Random mode](../openactive-integration-tests/README.md#userandomopportunities), it is important that any Opportunity is only used in one test. This increases test coverage by ensuring that tests use a variety of different Opportunities and, crucially, it prevents unexpected behaviours when tests are run in parallel.

Broker Microservice uses **Opportunity Locks** to ensure that an Opportunity is only used in one test. When an Opportunity is retrieved in Random mode, using [`POST /test-interface/datasets/:testDatasetIdentifier/opportunities`](#post-test-interfacedatasetstestdatasetidentifieropportunities), it is **locked**. Subsequent calls to get a random Opportunity with the same critera will not return the same Opportunity, as it is locked.

> In between runs of the tests in [Integration Tests](../openactive-integration-tests/), these locks need to be **released**, to ensure that all Opportunities are available for subsequent test runs.

Reviewer: ```suggestion
In between runs of the tests in [Integration Tests](../openactive-integration-tests/), these locks are **released**, to ensure that all Opportunities are available for subsequent test runs.
```

Reviewer: Just to clarify no additional action is required (as the release is automatic)
